overlord/snapshotstate/backend/backend.go

```diff
@@ -445,6 +446,149 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
+	}
+
+	// walk the cache directory to store the snapshot files from the
+	// temp import location to the real location
+	dir, err := osOpen(tempImportDir)
+	if err != nil {
+		return 0, nil, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	defer dir.Close()
+
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if len(names) > 0 {
+			// move the files into place with the new local set ID
+			newSnapNames, err := moveCachedSnapshots(tempImportDir, names, id)
+			// XXX: if we fail here we need to undo the snapshot
+			//      import
+			if err != nil {
+				return 0, nil, err
+			}
+			snapNames = append(snapNames, newSnapNames...)
+		}
+	}
+	if readErr != nil && readErr != io.EOF {
+		return 0, nil, fmt.Errorf("%s: failed read from import cache: %v", errPrefix, readErr)
+	}
+	return size, snapNames, err
+}
+
+func unpackVerifySnapshotImport(r io.Reader, p string) (exportFound bool, size int64, err error) {
+	tr := tar.NewReader(r)
+	var tarErr error
+	var header *tar.Header
+
+	var extractedSnapshots []string
+	for tarErr == nil {
+		header, tarErr = tr.Next()
+		if tarErr == io.EOF {
+			break
+		}
+		switch {
+		case tarErr != nil:
+			return false, 0, fmt.Errorf("failed reading snapshot import: %v", tarErr)
+		case header == nil:
+			// should not happen
+			return false, 0, fmt.Errorf("tar header not found")
+		case header.Typeflag == tar.TypeDir:
+			return false, 0, errors.New("unexpected directory in import file")
+		}
+
+		target := path.Join(p, header.Name)
+		if header.Name == "export.json" {
+			exportFound = true
+		} else {
+			extractedSnapshots = append(extractedSnapshots, target)
+		}
+
+		t, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+		if err != nil {
+			return false, 0, fmt.Errorf("failed creating file `%s`: %v", target, err)
+		}
+		defer t.Close()
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
+		}
+		size += writtenSize
+	}
+
+	// validate
+	for _, name := range extractedSnapshots {
+		// we don't care about set id at this point
+		r, err := backendOpen(name, ExtractFnameSetID)
+		if err != nil {
+			return false, 0, fmt.Errorf("cannot open snapshot: %v", err)
+		}
+		err = r.Check(context.TODO(), nil)
+		r.Close()
+		if err != nil {
+			return false, 0, fmt.Errorf("validation failed for %q: %v", name, err)
+		}
+	}
+	return exportFound, size, nil
+}
+
+func moveCachedSnapshots(cachedSnapshotsCDir string, names []string, id uint64) ([]string, error) {
+	snaps := []string{}
+	for _, name := range names {
+		if name == "export.json" {
+			// ignore metadata file
+			continue
+		}
+
+		old := path.Join(cachedSnapshotsCDir, name)
+
+		// read old snapshot, override its set id internally
+		r, err := backendOpen(old, id)
+		if err != nil {
+			return nil, fmt.Errorf("cannot open snapshot: %v", err)
+		}
+		r.Close()
+		if err != nil {
```

Comment:
We don't need this `if err != nil` now anymore AFAICT after the latest refactor that moves check out of this helper.

Reply:
I fixed this now